File: tools/cmd/vakt-ids/notify.go
package main

import (
	"fmt"
	"net"
	"os"
	"strings"
)

// notifySocketEnv is set by vakt-init for every service it supervises. When it
// is absent the daemon is being run by hand and there is nobody to tell.
const notifySocketEnv = "NOTIFY_SOCKET"

// notifyReady tells vakt-init the daemon has finished starting, so boot stops
// waiting on it and draws the panel.
//
// The protocol is newline-separated KEY=value pairs in a single Unix datagram,
// the same shape as sd_notify. Nothing here can fail in a way worth acting on:
// a daemon that cannot reach PID 1 still has a filesystem to watch, and
// refusing to work over a missed notification would turn a cosmetic problem
// into an outage.
func notifyReady(status string) {
	send("READY=1\nNAME=vakt-ids\nSTATUS=" + oneLine(status) + "\n")
}

func send(message string) {
	path := os.Getenv(notifySocketEnv)
	if path == "" {
		return
	}

	conn, err := net.Dial("unixgram", path)
	if err != nil {
		return
	}
	defer conn.Close()
	_, _ = conn.Write([]byte(message))
}

// oneLine flattens status text: a newline in a value would be read as the start
// of the next key.
func oneLine(text string) string {
	return strings.Join(strings.Fields(text), " ")
}

// readyStatus describes the baseline in the one line the panel has room for.
func readyStatus(fileCount int, roots []string) string {
	return fmt.Sprintf("watching %d file(s) under %s", fileCount, strings.Join(roots, ", "))
}

File: vakt-net/src/notify.rs
//! Telling vakt-init that the daemon has finished starting.
//!
//! PID 1 listens on the socket named by `NOTIFY_SOCKET` and holds boot until
//! the daemons check in. The protocol is newline-separated `KEY=value` pairs in
//! one datagram. Sending is best-effort: a daemon that cannot reach PID 1 has
//! nothing useful to do about it.

use std::os::unix::net::UnixDatagram;

/// Set by vakt-init for every service it supervises; absent when the daemon is
/// run by hand.
const SOCKET_ENV: &str = "NOTIFY_SOCKET";

/// Announces that the daemon is up. Safe to call more than once.
pub fn ready(status: &str) {
    send(&format!(
        "READY=1\nNAME=vakt-net\nSTATUS={}\n",
        one_line(status)
    ));
}

/// Updates the status line without claiming readiness.
pub fn status(status: &str) {
    send(&format!("NAME=vakt-net\nSTATUS={}\n", one_line(status)));
}

fn send(message: &str) {
    let Ok(path) = std::env::var(SOCKET_ENV) else {
        return;
    };
    let Ok(socket) = UnixDatagram::unbound() else {
        return;
    };
    let _ = socket.send_to(message.as_bytes(), path);
}

/// A newline in a value would be read as the start of the next key, so status
/// text is flattened before it goes on the wire.
fn one_line(text: &str) -> String {
    text.split_whitespace().collect::<Vec<_>>().join(" ")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn status_text_cannot_inject_another_field() {
        assert_eq!(
            one_line("connected\nSHUTDOWN=poweroff"),
            "connected SHUTDOWN=poweroff"
        );
    }

    #[test]
    fn whitespace_is_collapsed() {
        assert_eq!(
            one_line("  lease   acquired \t on eth0 "),
            "lease acquired on eth0"
        );
    }

    /// The daemon is runnable outside vakt-init, so no socket means no-op.
    #[test]
    fn sending_without_a_socket_is_harmless() {
        unsafe { std::env::remove_var(SOCKET_ENV) };
        ready("no listener");
        status("still no listener");
    }
}
